### PR TITLE
Do not overallocate SocketAddress.Buffer

### PR DESCRIPTION
**Pull request by antonfirsov**
We build [SocketAddress.cs](https://github.com/dotnet/runtime/blob/main/src/libraries/Common/src/System/Net/SocketAddress.cs) in two modes: to get the public type in System.Net.Primitives and - independently - as a System.Net.Sockets internal, so we can access internal members, and work withe the extra bytes allocated at the end of `SocketAddress.Buffer` for WSARecvFrom which needs a pinned pointer to the socket address size. This duplication was a refactoring technique when Framework code was ported to Core where System.Net.Sockets and System.Net.Primtives live in separate dll-s and socket code cannot access SocketAddress internals.

There seem to be two obvious problems:
1. Extra bytes are allocated in both compilation modes penalizing the public `SocketAddress` everywhere, including Unix platforms
2. The formula to calculate the number of bytes to be allocated is inherited from early .NET Framework days, and seems to be wrong, or at least overly defensive:

https://github.com/dotnet/runtime/blob/4f53c2f7e62df44f07cf410df8a0d439f42a0a71/src/libraries/Common/src/System/Net/SocketAddress.cs#L95

This **allocates 32 bytes for a 16-byte IPv4 address** for both `SocketAddress` and `Internals.SocketAddress` on a 64bit machine. ~The intention of the formula was probably to make sure that `lpFromlen`'s buffer is 4-byte aligned (assuming 32 bit), but I don't see any signs in [WSARecvFrom docs](https://learn.microsoft.com/en-us/windows/win32/api/winsock2/nf-winsock2-wsarecvfrom) that there is a need for any alignment.~ Edit: updated the code reintroducing alignment (though with a "smarter" formula), see discussion.

### CODE REVIEW

**Comment by wfurt** [comment]
The `lpFromlen` needs to be aligned to 32/64 boundary on ARM32/64. This is not requirement of the networking stack but more so limitation of the architecture that is not able to access misaligned int. (it it just less efficient on x86/x64)

**Comment by antonfirsov** [comment]
@wfurt I did some reading on this, and my understanding is that for windows API we need to comply with the [C standard](https://www.dii.uchile.cl/~daespino/files/Iso_C_1999_definition.pdf) which states in [3.2]:

> requirement that objects of a particular type be located on storage boundaries with addresses that are particular multiples of a byte address

Meaning that for `LPINT` [which is a pointer to a 32 byte `INT`](https://learn.microsoft.com/en-us/windows/win32/winprog/windows-data-types), the pointer needs to be 32-byte aligned, regardless of the platform.

ddb50fc updated the code to add the minimum necessary alignment.

File: src/libraries/Common/src/System/Net/SocketAddress.cs [hunk 1]
```diff
@@ -92,7 +92,15 @@ public SocketAddress(AddressFamily family, int size)
...
 
             InternalSize = size;
-            Buffer = new byte[(size / IntPtr.Size + 2) * IntPtr.Size];
+#if !SYSTEM_NET_PRIMITIVES_DLL && WINDOWS
+            // WSARecvFrom needs a pinned pointer to the 32bit socket address size: https://learn.microsoft.com/en-us/windows/win32/api/winsock2/nf-winsock2-wsarecvfrom
+            // Allocate sizeof(int) extra bytes at the end of Buffer ensuring sizeof(int) alignment for the LPINT lpFromlen pointer, so we don't need to pin anything else.
+            // The following forumla will extend 'size' to the 4-byte alignment boundary then add 4 more bytes.
+            // eg. size=16 will be extended to 20, while size=17 will be extended to 24.
+            const int PtrSize = sizeof(int);
+            size = (size + PtrSize -  1) / PtrSize * PtrSize + PtrSize;
```
Comment (antonfirsov) [inline src/libraries/Common/src/System/Net/SocketAddress.cs:101]:
```C#
const int PtrSize = sizeof(int);
int Extend(int size) => (size + PtrSize -  1) / PtrSize * PtrSize + PtrSize;

for (int i = 0; i < 20; i++) Console.Write($"{i,2} | ");
Console.WriteLine();
for (int i = 0; i < 20; i++) Console.Write($"{Extend(i),2} | ");
```

Will print:

```
0 |  1 |  2 |  3 |  4 |  5 |  6 |  7 |  8 |  9 | 10 | 11 | 12 | 13 | 14 | 15 | 16 | 17 | 18 | 19 |
4 |  8 |  8 |  8 |  8 | 12 | 12 | 12 | 12 | 16 | 16 | 16 | 16 | 20 | 20 | 20 | 20 | 24 | 24 | 24 |
```

File: src/libraries/Common/src/System/Net/SocketAddress.cs [hunk 2]
```diff
@@ -92,7 +92,13 @@ public SocketAddress(AddressFamily family, int size)
             }
 
             InternalSize = size;
-            Buffer = new byte[(size / IntPtr.Size + 2) * IntPtr.Size];
+#if !SYSTEM_NET_PRIMITIVES_DLL && WINDOWS
+            // WSARecvFrom needs a pinned pointer to the 32bit socket address size: https://learn.microsoft.com/en-us/windows/win32/api/winsock2/nf-winsock2-wsarecvfrom
+            // Allocate IntPtr.Size extra bytes at the end of Buffer ensuring IntPtr.Size alignment, so we don't need to pin anything else.
+            // The following forumla will extend 'size' to the alignment boundary then add IntPtr.Size more bytes.
+            size = (size + IntPtr.Size -  1) / IntPtr.Size * IntPtr.Size + IntPtr.Size;
```
Comment (antonfirsov) [inline src/libraries/Common/src/System/Net/SocketAddress.cs:99]:
```C#
int Extend(int size) => (size + IntPtr.Size - 1) / IntPtr.Size * IntPtr.Size + IntPtr.Size;

for (int i = 10; i < 30; i++) Console.Write($"{i,2} | ");
Console.WriteLine();
for (int i = 10; i < 30; i++) Console.Write($"{Extend(i),2} | ");
```

Will print:

```
10 | 11 | 12 | 13 | 14 | 15 | 16 | 17 | 18 | 19 | 20 | 21 | 22 | 23 | 24 | 25 | 26 | 27 | 28 | 29 |
24 | 24 | 24 | 24 | 24 | 24 | 24 | 32 | 32 | 32 | 32 | 32 | 32 | 32 | 32 | 40 | 40 | 40 | 40 | 40 |
```

Comment (wfurt) [reply src/libraries/Common/src/System/Net/SocketAddress.cs:99]: Right, the old code did
```
10 | 11 | 12 | 13 | 14 | 15 | 16 | 17 | 18 | 19 | 20 | 21 | 22 | 23 | 24 | 25 | 26 | 27 | 28 | 29 |
24 | 24 | 24 | 24 | 24 | 24 | 32 | 32 | 32 | 32 | 32 | 32 | 32 | 32 | 40 | 40 | 40 | 40 | 40 | 40 |
```

Sensible address structure would already be aligned IMHO and we really just need to put `int` on boundary of its size (e.g. 4) and we need to make sure it fits to the buffer. The new code does it and now we do it only for Windows where it is used.

File: src/libraries/Common/src/System/Net/SocketAddress.cs [hunk 3]
```diff
@@ -92,7 +92,13 @@ public SocketAddress(AddressFamily family, int size)
             }
 
             InternalSize = size;
-            Buffer = new byte[(size / IntPtr.Size + 2) * IntPtr.Size];
+#if !SYSTEM_NET_PRIMITIVES_DLL && WINDOWS
+            // WSARecvFrom needs a pinned pointer to the 32bit socket address size: https://learn.microsoft.com/en-us/windows/win32/api/winsock2/nf-winsock2-wsarecvfrom
+            // Allocate IntPtr.Size extra bytes at the end of Buffer ensuring IntPtr.Size alignment, so we don't need to pin anything else.
+            // The following forumla will extend 'size' to the alignment boundary then add IntPtr.Size more bytes.
```
Comment (wfurt) [inline src/libraries/Common/src/System/Net/SocketAddress.cs:98]: nit: forumla -> formula

**Review by wfurt — APPROVED** [review]
LGTM

**Comment by antonfirsov** [comment]
Outerloop failures are unrelated: #79731, #77414, #77012, #79820

**Pull request merged**